**Note: All experiments have been run on PL data only**

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from datetime import datetime
import warnings
import json
import operator
from matplotlib.ticker import FuncFormatter
from matplotlib.patches import Ellipse
import base64
from collections import defaultdict
import sys,os
import math
import random
import matplotlib.pylab as pyl
import itertools
import pickle
import swifter
import warnings
from unidecode import unidecode
from itertools import chain
from multiprocessing import  Pool

warnings.filterwarnings('ignore')

C:\Users\Anuj\Anaconda3\lib\site-packages\dask\dataframe\utils.py:15: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [ ]:
pd.set_option('max_colwidth', 999)

In [3]:
pd.set_option('display.max_columns', 1000)
pd.set_option("display.max_rows", 3000)
warnings.filterwarnings('ignore')

# Competitions 

- area: it denotes the geographic area associated with the league as a sub-document, using the ISO 3166-1 specification (https://www.iso.org/iso-3166-country-codes.html);
- format: the format of the competition. All competitions for clubs have value "Domestic league". The competitions for national teams have value "International cup";
- name: the official name of the competition (e.g., Italian first division, Spanish first division, World Cup, etc.);
- type: the typology of the competition. It is "club" for the competitions for clubs and "international" for the competitions for national teams (World Cup 2018, European Cup 2016);
- wyId: the unique identifier of the competition, assigned by Wyscout.

In [3]:
comps = pd.read_json('../data/competitions.json')
comps = pd.DataFrame(comps)
comps

name  wyId             format  \
0  Italian first division   524    Domestic league   
1  English first division   364    Domestic league   
2  Spanish first division   795    Domestic league   
3   French first division   412    Domestic league   
4   German first division   426    Domestic league   
5   European Championship   102  International cup   
6               World Cup    28  International cup   

                                                area           type  
0  {'name': 'Italy', 'id': '380', 'alpha3code': '...           club  
1  {'name': 'England', 'id': '0', 'alpha3code': '...           club  
2  {'name': 'Spain', 'id': '724', 'alpha3code': '...           club  
3  {'name': 'France', 'id': '250', 'alpha3code': ...           club  
4  {'name': 'Germany', 'id': '276', 'alpha3code':...           club  
5  {'name': '', 'id': 0, 'alpha3code': 'XEU', 'al...  international  
6  {'name': '', 'id': 0, 'alpha3code': 'XWO', 'al...  international

# Matches 

- competitionId: the identifier of the competition to which the match belongs to. It is a integer and refers to the field "wyId" of the competition document;
- date and dateutc: the former specifies date and time when the match starts in explicit format (e.g., May 20, 2018 at 8:45:00 PM GMT+2), the latter contains the same information but in the compact format YYYY-MM-DD hh:mm:ss;
- duration: the duration of the match. It can be "Regular" (matches of regular duration of 90 minutes + stoppage time), "ExtraTime" (matches with supplementary times, as it may happen for matches in continental or international competitions), or "Penalities" (matches which end at penalty kicks, as it may happen for continental or international competitions);
- gameweek: the week of the league, starting from the beginning of the league;
- label: contains the name of the two clubs and the result of the match (e.g., "Lazio - Internazionale, 2 - 3");
- roundID: indicates the match-day of the competition to which the match belongs to. During a competition for soccer clubs, each of the participating clubs plays against each of the other clubs twice, once at home and once away. The matches are organized in match-days: all the matches in match-day i are played before the matches in match-day i + 1, even tough some matches can be anticipated or postponed to facilitate players and clubs participating in Continental or Intercontinental competitions. During a competition for national teams, the "roundID" indicates the stage of the competition (eliminatory round, round of 16, quarter finals, semifinals, final);
- seasonId: indicates the season of the match;
- status: it can be "Played" (the match has officially finished), "Cancelled" (the match has been canceled for some reason), "Postponed" (the match has been postponed and no new date and time is available yet) or "Suspended" (the match has been suspended and no new date and time is available yet);
- venue: the stadium where the match was held (e.g., "Stadio Olimpico");
- winner: the identifier of the team which won the game, or 0 if the match ended with a draw;
- wyId: the identifier of the match, assigned by Wyscout;
- teamsData: it contains several subfields describing information about each team that is playing that match: such as lineup, bench composition, list of substitutions, coach and scores:
- hasFormation: it has value 0 if no formation (lineups and benches) is present, and 1 otherwise;
- score: the number of goals scored by the team during the match (not counting penalties);
- scoreET: the number of goals scored by the team during the match, including the extra time (not counting penalties);
- scoreHT: the number of goals scored by the team during the first half of the match;
- scoreP: the total number of goals scored by the team after the penalties;
- side: the team side in the match (it can be "home" or "away");
- teamId: the identifier of the team;
- coachId: the identifier of the team's coach;
- bench: the list of the team's players that started the match in the bench and some basic statistics about their performance during the match (goals, own goals, cards);
- lineup: the list of the team's players in the starting lineup and some basic statistics about their performance during the match (goals, own goals, cards);
- substitutions: the list of team's substitutions during the match, describing the players involved and the minute of the substitution.

In [4]:
matches = pd.read_json('../data/matches/matches_England.json')
matches = pd.DataFrame(matches)
print(matches.shape)
display(matches)

(380, 14)


status  roundId  gameweek  \
0    Played  4405654        38   
1    Played  4405654        38   
2    Played  4405654        38   
3    Played  4405654        38   
4    Played  4405654        38   
5    Played  4405654        38   
6    Played  4405654        38   
7    Played  4405654        38   
8    Played  4405654        38   
9    Played  4405654        38   
10   Played  4405654        31   
11   Played  4405654        31   
12   Played  4405654        31   
13   Played  4405654        31   
14   Played  4405654        35   
15   Played  4405654        31   
16   Played  4405654        37   
17   Played  4405654        37   
18   Played  4405654        37   
19   Played  4405654        37   
20   Played  4405654        37   
21   Played  4405654        37   
22   Played  4405654        37   
23   Played  4405654        37   
24   Played  4405654        37   
25   Played  4405654        37   
26   Played  4405654        36   
27   Played  4405654        36   
28   Played  4405654        36   
29   Played  4405654        36   
30   Played  4405654        36   
31   Played  4405654        36   
32   Played  4405654        36   
33   Played  4405654        36   
34   Played  4405654        36   
35   Played  4405654        36   
36   Played  4405654        35   
37   Played  4405654        35   
38   Played  4405654        35   
39   Played  4405654        35   
40   Played  4405654        35   
41   Played  4405654        35   
42   Played  4405654        31   
43   Played  4405654        35   
44   Played  4405654        35   
45   Played  4405654        35   
46   Played  4405654        34   
47   Played  4405654        34   
48   Played  4405654        34   
49   Played  4405654        34   
50   Played  4405654        34   
51   Played  4405654        34   
52   Played  4405654        34   
53   Played  4405654        34   
54   Played  4405654        34   
55   Played  4405654        34   
56   Played  4405654        33   
57   Played  4405654        33   
58   Played  4405654        33   
59   Played  4405654        33   
60   Played  4405654        33   
61   Played  4405654        33   
62   Played  4405654        33   
63   Played  4405654        33   
64   Played  4405654        33   
65   Played  4405654        33   
66   Played  4405654        32   
67   Played  4405654        32   
68   Played  4405654        32   
69   Played  4405654        32   
70   Played  4405654        32   
71   Played  4405654        32   
72   Played  4405654        32   
73   Played  4405654        32   
74   Played  4405654        32   
75   Played  4405654        32   
76   Played  4405654        31   
77   Played  4405654        31   
78   Played  4405654        31   
79   Played  4405654        31   
80   Played  4405654        30   
81   Played  4405654        30   
82   Played  4405654        30   
83   Played  4405654        30   
84   Played  4405654        30   
85   Played  4405654        30   
86   Played  4405654        30   
87   Played  4405654        30   
88   Played  4405654        30   
89   Played  4405654        30   
90   Played  4405654        29   
91   Played  4405654        29   
92   Played  4405654        29   
93   Played  4405654        29   
94   Played  4405654        29   
95   Played  4405654        29   
96   Played  4405654        29   
97   Played  4405654        29   
98   Played  4405654        29   
99   Played  4405654        29   
100  Played  4405654        28   
101  Played  4405654        28   
102  Played  4405654        28   
103  Played  4405654        28   
104  Played  4405654        28   
105  Played  4405654        28   
106  Played  4405654        28   
107  Played  4405654        28   
108  Played  4405654        28   
109  Played  4405654        28   
110  Played  4405654        27   
111  Played  4405654        27   
112  Played  4405654        27   
113  Played  4405654        27   
114  Played  4405654        27   
115  Played  4405654        27   
116  Played  4405654       

Sample 'TeamsData' row entry:

# Events

- eventId: the identifier of the event's type. Each eventId is associated with an event name (see next point);
- eventName: tteamIdhe name of the event's type. There are seven types of events: pass, foul, shot, duel, free kick, offside and touch;
- subEventId: the identifier of the subevent's type. Each subEventId is associated with a subevent name (see next point);
- subEventName: the name of the subevent's type. Each event type is associated with a different set of subevent types;
- tags: a list of event tags, each one describes additional information about the event (e.g., accurate). Each event type is associated with a different set of tags;
- eventSec: the time when the event occurs (in seconds since the beginning of the current half of the match);
- id: a unique identifier of the event;
- matchId: the identifier of the match the event refers to. The identifier refers to the field "wyId" in the match dataset;
- matchPeriod: the period of the match. It can be "1H" (first half of the match), "2H" (second half of the match), "E1" (first extra time), "E2" (second extra time) or "P" (penalties time);
- playerId: the identifier of the player who generated the event. The identifier refers to the field "wyId" in a player dataset;
- positions: the origin and destination positions associated with the event. Each position is a pair of coordinates (x, y). The x and y coordinates are always in the range [0, 100] and indicate the percentage of the field from the perspective of the attacking team. In particular, the value of the x coordinate indicates the event's nearness (in percentage) to the opponent's goal, while the value of the y coordinates indicates the event's nearness (in percentage) to the right side of the field;
- teamId: the identifier of the player's team. The identifier refers to the field "wyId" in the team dataset.

In [5]:
events = pd.read_json('../data/events/events_England.json')
events = pd.DataFrame(events)
print(events.shape)
display(events)

(643150, 12)


eventId           subEventName  \
0             8            Simple pass   
1             8              High pass   
2             8              Head pass   
3             8              Head pass   
4             8            Simple pass   
...         ...                    ...   
643145        5  Ball out of the field   
643146        3                 Corner   
643147        1               Air duel   
643148        1               Air duel   
643149       10                   Shot   

                                                     tags  playerId  \
0                                          [{'id': 1801}]     25413   
1                                          [{'id': 1801}]    370224   
2                                          [{'id': 1801}]      3319   
3                                          [{'id': 1801}]    120339   
4                                          [{'id': 1801}]    167145   
...                                                   ...       ...   
643145                                                 []         0   
643146           [{'id': 302}, {'id': 801}, {'id': 1801}]     70965   
643147                        [{'id': 701}, {'id': 1802}]      7919   
643148                        [{'id': 703}, {'id': 1801}]      8005   
643149  [{'id': 403}, {'id': 201}, {'id': 1216}, {'id'...      8005   

                                         positions  matchId     eventName  \
0         [{'y': 49, 'x': 49}, {'y': 78, 'x': 31}]  2499719          Pass   
1         [{'y': 78, 'x': 31}, {'y': 75, 'x': 51}]  2499719          Pass   
2         [{'y': 75, 'x': 51}, {'y': 71, 'x': 35}]  2499719          Pass   
3         [{'y': 71, 'x': 35}, {'y': 95, 'x': 41}]  2499719          Pass   
4         [{'y': 95, 'x': 41}, {'y': 88, 'x': 72}]  2499719          Pass   
...                                            ...      ...           ...   
643145   [{'y': 32, 'x': 0}, {'y': 100, 'x': 100}]  2500098  Interruption   
643146  [{'y': 100, 'x': 100}, {'y': 47, 'x': 88}]  2500098     Free Kick   
643147    [{'y': 53, 'x': 12}, {'y': 50, 'x': 14}]  2500098          Duel   
643148    [{'y': 47, 'x': 88}, {'y': 50, 'x': 86}]  2500098          Duel   
643149      [{'y': 50, 'x': 86}, {'y': 0, 'x': 0}]  2500098          Shot   

        teamId matchPeriod     eventSec subEventId         id  
0         1609          1H     2.758649         85  177959171  
1         1609          1H     4.946850         83  177959172  
2         1609          1H     6.542188         82  177959173  
3         1609          1H     8.143395         82  177959174  
4         1609          1H    10.302366         85  177959175  
...        ...         ...          ...        ...        ...  
643145    1623          2H  2796.732525         50  251596409  
643146    1633          2H  2829.821084         30  251596232  
643147    1623          2H  2831.211419         10  251596410  
643148    1633          2H  2832.434399         10  251596234  
643149    1633          2H  2834.127168        100  251596236  

[643150 rows x 12 columns]

In [6]:
print(len(events.matchId.unique()))
print(matches.shape)

380
(380, 14)


In [7]:
events.matchId.value_counts().mean()

1692.5

Approximately, every 3.19 seconds, an event activity is registered

# Player 

- birthArea: geographic information about the player's birth area;
- birthDate: the birth date of the player, in the format "YYYY-MM-DD";
- currentNationalTeamId: the identifier of the national team where the players currently plays;
- currentTeamId: the identifier of the team where the player plays for. The identifier refers to the field "wyId" in a team document;
- firstName: the first name of the player;
- lastName: the last name of the player;
- foot: the preferred foot of the player;
- height: the height of the player (in centimeters);
- middleName: the middle name (if any) of the player;
- passportArea: the geographic area associated with the player's current passport;
- role: the main role of the player. It is a subdocument containing the role's name and two abbreviations of it;
- shortName2: the short name of the player;
- weight: the weight of the player (in kilograms);
- wyId: the identifier of the player, assigned by Wyscout.

In [43]:
players = pd.read_json('../data/players.json')
players = pd.DataFrame(players)
print(players.shape)
display(players)

(3603, 14)


passportArea  weight  \
0     {'name': 'Turkey', 'id': '792', 'alpha3code': ...      78   
1     {'name': 'Senegal', 'id': '686', 'alpha3code':...      73   
2     {'name': 'France', 'id': '250', 'alpha3code': ...      72   
3     {'name': 'Senegal', 'id': '686', 'alpha3code':...      82   
4     {'name': 'France', 'id': '250', 'alpha3code': ...      84   
...                                                 ...     ...   
3598  {'name': 'Tunisia', 'id': 788, 'alpha3code': '...      72   
3599  {'name': 'Peru', 'id': 604, 'alpha3code': 'PER...      76   
3600  {'name': 'Peru', 'id': 604, 'alpha3code': 'PER...      78   
3601  {'name': 'Morocco', 'id': 504, 'alpha3code': '...      70   
3602  {'name': 'Panama', 'id': 591, 'alpha3code': 'P...       0   

              firstName middleName             lastName currentTeamId  \
0                 Harun                           Tekin          4502   
1                Malang                            Sarr          3775   
2                  Over                        Mandanda          3772   
3     Alfred John Momar                         N'Diaye           683   
4              Ibrahima                     Konat\u00e9          2975   
...                 ...        ...                  ...           ...   
3598                Ali                    Ma\u00e2loul         16041   
3599     Carlos Alberto             C\u00e1ceda Oyaguez         15591   
3600  Miguel Gianpierre                   Araujo Blanco         12072   
3601         Ahmed Reda                       Tagnaouti         16183   
3602            Ricardo                   Guardia Avila         62943   

       birthDate  height                                               role  \
0     1989-06-17     187  {'code2': 'GK', 'code3': 'GKP', 'name': 'Goalk...   
1     1999-01-23     182  {'code2': 'DF', 'code3': 'DEF', 'name': 'Defen...   
2     1998-10-26     176  {'code2': 'GK', 'code3': 'GKP', 'name': 'Goalk...   
3     1990-03-06     187  {'code2': 'MD', 'code3': 'MID', 'name': 'Midfi...   
4     1999-05-25     192  {'code2': 'DF', 'code3': 'DEF', 'name': 'Defen...   
...          ...     ...                                                ...   
3598  1990-01-01     175  {'code2': 'DF', 'code3': 'DEF', 'name': 'Defen...   
3599  1991-09-27     183  {'code2': 'GK', 'code3': 'GKP', 'name': 'Goalk...   
3600  1994-10-24     179  {'code2': 'DF', 'code3': 'DEF', 'name': 'Defen...   
3601  1996-04-05     182  {'code2': 'GK', 'code3': 'GKP', 'name': 'Goalk...   
3602  1997-02-04       0  {'code2': 'MD', 'code3': 'MID', 'name': 'Midfi...   

                                              birthArea    wyId   foot  \
0     {'name': 'Turkey', 'id': '792', 'alpha3code': ...   32777  right   
1     {'name': 'France', 'id': '250', 'alpha3code': ...  393228   left   
2     {'name': 'France', 'id': '250', 'alpha3code': ...  393230          
3     {'name': 'France', 'id': '250', 'alpha3code': ...   32793  right   
4     {'name': 'France', 'id': '250', 'alpha3code': ...  393247  right   
...                                                 ...     ...    ...   
3598  {'name': 'Tunisia', 'id': 788, 'alpha3code': '...  120839   left   
3599  {'name': 'Peru', 'id': 604, 'alpha3code': 'PER...  114736  right   
3600  {'name': 'Peru', 'id': 604, 'alpha3code': 'PER...  114908  right   
3601  {'name': 'Morocco', 'id': 504, 'alpha3code': '...  285583  right   
3602  {'name': 'Panama', 'id': 591, 'alpha3code': 'P...  361536   left   

            shortName currentNationalTeamId  
0            H. Tekin                  4687  
1             M. Sarr                  4423  
2         O. Mandanda                  null  
3          A. N'Diaye                 19314  
4      I. Konat\u00e9                  null  
...               ...                   ...  
3598  A. Ma\u00e2loul                  null  
3599   C. C\u00e1ceda                  null  
3600        M. Araujo                  null  
3601     A. Tagnaouti                  null  
3602         R. A

# Events + Player DF 

In [7]:
events_com = pd.merge(right=players, left=events, right_on='wyId', left_on='playerId', how='left')
print(events_com.shape)
display(events_com)

(643150, 26)


eventId           subEventName  \
0             8            Simple pass   
1             8              High pass   
2             8              Head pass   
3             8              Head pass   
4             8            Simple pass   
...         ...                    ...   
643145        5  Ball out of the field   
643146        3                 Corner   
643147        1               Air duel   
643148        1               Air duel   
643149       10                   Shot   

                                                     tags  playerId  \
0                                          [{'id': 1801}]     25413   
1                                          [{'id': 1801}]    370224   
2                                          [{'id': 1801}]      3319   
3                                          [{'id': 1801}]    120339   
4                                          [{'id': 1801}]    167145   
...                                                   ...       ...   
643145                                                 []         0   
643146           [{'id': 302}, {'id': 801}, {'id': 1801}]     70965   
643147                        [{'id': 701}, {'id': 1802}]      7919   
643148                        [{'id': 703}, {'id': 1801}]      8005   
643149  [{'id': 403}, {'id': 201}, {'id': 1216}, {'id'...      8005   

                                         positions  matchId     eventName  \
0         [{'y': 49, 'x': 49}, {'y': 78, 'x': 31}]  2499719          Pass   
1         [{'y': 78, 'x': 31}, {'y': 75, 'x': 51}]  2499719          Pass   
2         [{'y': 75, 'x': 51}, {'y': 71, 'x': 35}]  2499719          Pass   
3         [{'y': 71, 'x': 35}, {'y': 95, 'x': 41}]  2499719          Pass   
4         [{'y': 95, 'x': 41}, {'y': 88, 'x': 72}]  2499719          Pass   
...                                            ...      ...           ...   
643145   [{'y': 32, 'x': 0}, {'y': 100, 'x': 100}]  2500098  Interruption   
643146  [{'y': 100, 'x': 100}, {'y': 47, 'x': 88}]  2500098     Free Kick   
643147    [{'y': 53, 'x': 12}, {'y': 50, 'x': 14}]  2500098          Duel   
643148    [{'y': 47, 'x': 88}, {'y': 50, 'x': 86}]  2500098          Duel   
643149      [{'y': 50, 'x': 86}, {'y': 0, 'x': 0}]  2500098          Shot   

        teamId matchPeriod     eventSec subEventId         id  \
0         1609          1H     2.758649         85  177959171   
1         1609          1H     4.946850         83  177959172   
2         1609          1H     6.542188         82  177959173   
3         1609          1H     8.143395         82  177959174   
4         1609          1H    10.302366         85  177959175   
...        ...         ...          ...        ...        ...   
643145    1623          2H  2796.732525         50  251596409   
643146    1633          2H  2829.821084         30  251596232   
643147    1623          2H  2831.211419         10  251596410   
643148    1633          2H  2832.434399         10  251596234   
643149    1633          2H  2834.127168        100  251596236   

                                             passportArea  weight  \
0       {'name': 'Guadeloupe', 'id': 312, 'alpha3code'...    73.0   
1       {'name': 'England', 'id': 0, 'alpha3code': 'XE...    75.0   
2       {'name': 'Germany', 'id': '276', 'alpha3code':...    76.0   
3       {'name': 'Egypt', 'id': '818', 'alpha3code': '...    70.0   
4       {'name': 'Spain', 'id': 724, 'alpha3code': 'ES...    74.0   
...                                                   ...     ...   
643145                                                NaN     NaN   
643146  {'name': 'Angola', 'id': '24', 'alpha3code': '...    67.0   
643147  {'name': 'England', 'id': 0, 'alpha3code': 'XE...    82.0   
643148  {'name': 'Wales', 'id': 0, 'alpha3code': 'XWA'...    90.0   
643149  {'name': 'Wales', 'id': 0, 'alpha3code': 'XWA'...    90.0   

                   firstName middleName                 lastName  \
0                  Alexandre                   

In [8]:
#Those playerIds from df events that dont have an entry in df players
noplayer = events_com[events_com['firstName'].isna().values]
noplayer

eventId            subEventName  \
8             1                Air duel   
23            1  Ground loose ball duel   
36            3               Free Kick   
41            1   Ground defending duel   
54            1   Ground attacking duel   
...         ...                     ...   
643080        5   Ball out of the field   
643101        3               Free Kick   
643117        1   Ground defending duel   
643140        3               Free Kick   
643145        5   Ball out of the field   

                                            tags  playerId  \
8                    [{'id': 703}, {'id': 1801}]         0   
23                   [{'id': 701}, {'id': 1802}]         0   
36                                [{'id': 1801}]         0   
41      [{'id': 502}, {'id': 701}, {'id': 1802}]         0   
54                   [{'id': 701}, {'id': 1802}]         0   
...                                          ...       ...   
643080                                        []         0   
643101                            [{'id': 1801}]         0   
643117  [{'id': 502}, {'id': 701}, {'id': 1802}]         0   
643140                            [{'id': 1801}]         0   
643145                                        []         0   

                                          positions  matchId     eventName  \
8          [{'y': 85, 'x': 61}, {'y': 80, 'x': 67}]  2499719          Duel   
23         [{'y': 78, 'x': 33}, {'y': 85, 'x': 29}]  2499719          Duel   
36         [{'y': 44, 'x': 65}, {'y': 31, 'x': 65}]  2499719     Free Kick   
41          [{'y': 86, 'x': 4}, {'y': 97, 'x': 10}]  2499719          Duel   
54         [{'y': 27, 'x': 89}, {'y': 27, 'x': 95}]  2499719          Duel   
...                                             ...      ...           ...   
643080  [{'y': 100, 'x': 51}, {'y': 100, 'x': 100}]  2500098  Interruption   
643101     [{'y': 87, 'x': 40}, {'y': 80, 'x': 56}]  2500098     Free Kick   
643117       [{'y': 34, 'x': 7}, {'y': 28, 'x': 5}]  2500098          Duel   
643140     [{'y': 81, 'x': 70}, {'y': 70, 'x': 59}]  2500098     Free Kick   
643145    [{'y': 32, 'x': 0}, {'y': 100, 'x': 100}]  2500098  Interruption   

        teamId matchPeriod     eventSec subEventId         id passportArea  \
8         1609          1H    14.765321         10  177961218          NaN   
23        1609          1H    38.191874         13  177961219          NaN   
36        1609          1H    74.737141         31  177959198          NaN   
41        1631          1H    85.074043         12  177961220          NaN   
54        1631          1H   152.127886         11  177961221          NaN   
...        ...         ...          ...        ...        ...          ...   
643080    1623          2H  2589.146520         50  251596374          NaN   
643101    1623          2H  2665.580593         31  251596391          NaN   
643117    1623          2H  2712.941293         12  251596987          NaN   
643140    1633          2H  2780.300522         31  251596224          NaN   
643145    1623          2H  2796.732525         50  251596409          NaN   

        weight firstName middleName lastName currentTeamId birthDate  height  \
8          NaN       NaN        NaN      NaN           NaN       NaN     NaN   
23         NaN       NaN        NaN      NaN           NaN       NaN     NaN   
36         NaN       NaN        NaN      NaN           NaN       NaN     NaN   
41         NaN       NaN        NaN      NaN           NaN       NaN     NaN   
54         NaN       NaN        NaN      NaN           NaN       NaN     NaN   
...        ...       ...        ...      ...           ...       ...     ...   
643080     NaN       NaN        NaN      NaN           NaN       NaN     NaN   
643101     NaN       NaN        NaN      NaN           NaN       NaN     NaN   
643117     NaN       NaN        NaN      NaN           NaN       NaN     NaN   
643140     NaN       NaN        NaN      NaN           NaN       NaN    

In [64]:
noplayer.subEventName.value_counts()

Ball out of the field     27326
Ground defending duel      6206
Ground attacking duel      4841
Ground loose ball duel     4096
Air duel                   2972
Free Kick                  1550
Throw in                    652
Whistle                     201
Foul                         59
Goal kick                    46
Touch                        40
Corner                       28
Launch                        3
Simple pass                   3
                              3
Shot                          1
Hand foul                     1
Reflexes                      1
Free kick cross               1
Clearance                     1
Name: subEventName, dtype: int64

These entries are the interruptions that dont need a player (mostly, rest are data collection anomalies)

***Q: 'tags' values meaning in df events?***
 [describes additional information about the event (e.g., accurate)]

In [ ]:
events_com = events_com.dropna()
events_com = events_com.reset_index(drop=True)

# Extracting Player Position from Match Lineup 

## Extracting Defender Position 

In [14]:
defence = pd.read_csv('../data/PL1718Defense.csv')

In [ ]:
defence.drop(columns = ['Unnamed: 0'], inplace=True)

In [16]:
def clean_tokenize(s):
    s = s.replace(",", ' ')
    s = s.replace("[", '')
    s = s.replace("]", '')
    s = s.replace("'", '')
    s = s.split()
    return s

In [ ]:
defence['home_team_defense'] = defence['home_team_defense'].apply(lambda x: clean_tokenize(x))
defence['away_team_defense'] = defence['away_team_defense'].apply(lambda x: clean_tokenize(x))

In [27]:
def add_home_posn(df):
    for i in range(0, len(df)):
        if(len(df['home_team_defense'][i])==4):
            df['home_RB'][i] = df['home_team_defense'][i][0]
            df['home_R-CB'][i] = df['home_team_defense'][i][1]
            df['home_L-CB'][i] = df['home_team_defense'][i][2]
            df['home_LB'][i] = df['home_team_defense'][i][3]
            df['home_backline'][i] = 4
        elif(len(df['home_team_defense'][i])==3):
            df['home_RCB'][i] = df['home_team_defense'][i][0]
            df['home_CB'][i] = df['home_team_defense'][i][1]
            df['home_LCB'][i] = df['home_team_defense'][i][2]
            df['home_backline'][i] = 3
        else:
            df['home_RWB'][i] = df['home_team_defense'][i][0]
            df['home_RCB'][i] = df['home_team_defense'][i][1]
            df['home_CB'][i] = df['home_team_defense'][i][2]
            df['home_LCB'][i] = df['home_team_defense'][i][3]
            df['home_LWB'][i] = df['home_team_defense'][i][4]
            df['home_backline'][i] = 5

In [29]:
def add_away_posn(df):
    for i in range(0, len(df)):
        if(len(df['away_team_defense'][i])==4):
            df['away_RB'][i] = df['away_team_defense'][i][0]
            df['away_R-CB'][i] = df['away_team_defense'][i][1]
            df['away_L-CB'][i] = df['away_team_defense'][i][2]
            df['away_LB'][i] = df['away_team_defense'][i][3]
            df['away_backline'][i] = 4
        elif(len(df['away_team_defense'][i])==3):
            df['away_RCB'][i] = df['away_team_defense'][i][0]
            df['away_CB'][i] = df['away_team_defense'][i][1]
            df['away_LCB'][i] = df['away_team_defense'][i][2]
            df['away_backline'][i] = 3
        else:
            df['away_RWB'][i] = df['away_team_defense'][i][0]
            df['away_RCB'][i] = df['away_team_defense'][i][1]
            df['away_CB'][i] = df['away_team_defense'][i][2]
            df['away_LCB'][i] = df['away_team_defense'][i][3]
            df['away_LWB'][i] = df['away_team_defense'][i][4]
            df['away_backline'][i] = 5

In [20]:
defence = defence.reindex(columns=defence.columns.tolist() + [
    'home_RB',
    'home_R-CB',
    'home_L-CB',
    'home_LB',
    'home_RCB',
    'home_CB',
    'home_LCB',
    'home_RWB',
    'home_LWB',
    'away_RB',
    'away_R-CB',
    'away_L-CB',
    'away_LB',
    'away_RCB',
    'away_CB',
    'away_LCB',
    'away_RWB',
    'away_LWB',
    'home_backline',
    'away_backline'
])

In [30]:
add_home_posn(defence)
add_away_posn(defence)

In [31]:
defence

match_id       home_team       away_team  match_date  \
0       22712         Burnley     Bournemouth  2018-05-13   
1       22713  Crystal Palace       West Brom  2018-05-13   
2       22714    Huddersfield         Arsenal  2018-05-13   
3       22715       Liverpool        Brighton  2018-05-13   
4       22716         Man Utd         Watford  2018-05-13   
5       22717       Newcastle         Chelsea  2018-05-13   
6       22718     Southampton        Man City  2018-05-13   
7       22719         Swansea           Stoke  2018-05-13   
8       22720           Spurs       Leicester  2018-05-13   
9       22721        West Ham         Everton  2018-05-13   
10      22651        West Ham         Man Utd  2018-05-04   
11      22685         Chelsea    Huddersfield  2018-05-04   
12      22645       Leicester         Arsenal  2018-05-04   
13      22647        Man City        Brighton  2018-05-04   
14      22650           Spurs       Newcastle  2018-05-04   
15      22649         Swansea     Southampton  2018-05-04   
16      22708        Man City    Huddersfield  2018-05-04   
17      22702         Arsenal         Burnley  2018-05-04   
18      22705         Chelsea       Liverpool  2018-05-04   
19      22704        Brighton         Man Utd  2018-05-04   
20      22709           Stoke  Crystal Palace  2018-05-04   
21      22703     Bournemouth         Swansea  2018-05-04   
22      22707       Leicester        West Ham  2018-05-04   
23      22710         Watford       Newcastle  2018-05-04   
24      22711       West Brom           Spurs  2018-05-04   
25      22706         Everton     Southampton  2018-05-04   
26      22700           Spurs         Watford  2018-04-28   
27      22701        West Ham        Man City  2018-04-28   
28      22696         Man Utd         Arsenal  2018-04-28   
29      22695       Liverpool           Stoke  2018-04-28   
30      22692         Burnley        Brighton  2018-04-28   
31      22693  Crystal Palace       Leicester  2018-04-28   
32      22694    Huddersfield         Everton  2018-04-28   
33      22697       Newcastle       West Brom  2018-04-28   
34      22698     Southampton     Bournemouth  2018-04-28   
35      22699         Swansea         Chelsea  2018-04-28   
36      22686         Everton       Newcastle  2018-04-21   
37      22682         Arsenal        West Ham  2018-04-21   
38      22689           Stoke         Burnley  2018-04-21   
39      22688        Man City         Swansea  2018-04-21   
40      22691       West Brom       Liverpool  2018-04-21   
41      22690         Watford  Crystal Palace  2018-04-21   
42      22643         Burnley         Chelsea  2018-04-14   
43      22687       Leicester     Southampton  2018-04-14   
44      22683     Bournemouth         Man Utd  2018-04-14   
45      22684        Brighton           Spurs  2018-04-14   
46      22681        West Ham           Stoke  2018-04-14   
47      22680           Spurs        Man City  2018-04-14   
48      22677       Newcastle         Arsenal  2018-04-14   
49      22676         Man Utd       West Brom  2018-04-14   
50      22678     Southampton         Chelsea  2018-04-14   
51      22672         Burnley       Leicester  2018-04-14   
52      22673  Crystal Palace        Brighton  2018-04-14   
53      22674    Huddersfield         Watford  2018-04-14   
54      22679         Swansea         Everton  2018-04-14   
55      22675       Liverpool     Bournemouth  2018-04-14   
56      22662         Arsenal     Southampton  2018-04-07   
57      22665         Chelsea        West Ham  2018-04-07   
58      22666         Everton       Liverpool  2018-04-07   
59      22663     Bournemouth  Crystal Palace  2018-04-07   
60      22664        Brighton    Huddersfield  2018-04-07   
61      22667       Leicester       Newcastle  2018-04-07   
62      22669           Stoke           Spurs  2018-04-07   
63      22670         Watford         Burnley  2018-04-07   
64      22671       West Brom         Sw

## Merge Match_IDs 

In [32]:
def label_splitter(df):
    df['label'] = df['label'].apply(lambda x: x.replace(",", ' '))
    df['label'] = df['label'].apply(lambda x: x.split('  '))
    df['score'] = df['label'].apply(lambda x: x[1])
    df['label'] = df['label'].apply(lambda x: x[0].split(' - '))
    df['home_team'] = df['label'].apply(lambda x: x[0])
    df['away_team'] = df['label'].apply(lambda x: x[1])
    return df

In [33]:
label_splitter(matches)

status  roundId  gameweek  \
0    Played  4405654        38   
1    Played  4405654        38   
2    Played  4405654        38   
3    Played  4405654        38   
4    Played  4405654        38   
5    Played  4405654        38   
6    Played  4405654        38   
7    Played  4405654        38   
8    Played  4405654        38   
9    Played  4405654        38   
10   Played  4405654        31   
11   Played  4405654        31   
12   Played  4405654        31   
13   Played  4405654        31   
14   Played  4405654        35   
15   Played  4405654        31   
16   Played  4405654        37   
17   Played  4405654        37   
18   Played  4405654        37   
19   Played  4405654        37   
20   Played  4405654        37   
21   Played  4405654        37   
22   Played  4405654        37   
23   Played  4405654        37   
24   Played  4405654        37   
25   Played  4405654        37   
26   Played  4405654        36   
27   Played  4405654        36   
28   Played  4405654        36   
29   Played  4405654        36   
30   Played  4405654        36   
31   Played  4405654        36   
32   Played  4405654        36   
33   Played  4405654        36   
34   Played  4405654        36   
35   Played  4405654        36   
36   Played  4405654        35   
37   Played  4405654        35   
38   Played  4405654        35   
39   Played  4405654        35   
40   Played  4405654        35   
41   Played  4405654        35   
42   Played  4405654        31   
43   Played  4405654        35   
44   Played  4405654        35   
45   Played  4405654        35   
46   Played  4405654        34   
47   Played  4405654        34   
48   Played  4405654        34   
49   Played  4405654        34   
50   Played  4405654        34   
51   Played  4405654        34   
52   Played  4405654        34   
53   Played  4405654        34   
54   Played  4405654        34   
55   Played  4405654        34   
56   Played  4405654        33   
57   Played  4405654        33   
58   Played  4405654        33   
59   Played  4405654        33   
60   Played  4405654        33   
61   Played  4405654        33   
62   Played  4405654        33   
63   Played  4405654        33   
64   Played  4405654        33   
65   Played  4405654        33   
66   Played  4405654        32   
67   Played  4405654        32   
68   Played  4405654        32   
69   Played  4405654        32   
70   Played  4405654        32   
71   Played  4405654        32   
72   Played  4405654        32   
73   Played  4405654        32   
74   Played  4405654        32   
75   Played  4405654        32   
76   Played  4405654        31   
77   Played  4405654        31   
78   Played  4405654        31   
79   Played  4405654        31   
80   Played  4405654        30   
81   Played  4405654        30   
82   Played  4405654        30   
83   Played  4405654        30   
84   Played  4405654        30   
85   Played  4405654        30   
86   Played  4405654        30   
87   Played  4405654        30   
88   Played  4405654        30   
89   Played  4405654        30   
90   Played  4405654        29   
91   Played  4405654        29   
92   Played  4405654        29   
93   Played  4405654        29   
94   Played  4405654        29   
95   Played  4405654        29   
96   Played  4405654        29   
97   Played  4405654        29   
98   Played  4405654        29   
99   Played  4405654        29   
100  Played  4405654        28   
101  Played  4405654        28   
102  Played  4405654        28   
103  Played  4405654        28   
104  Played  4405654        28   
105  Played  4405654        28   
106  Played  4405654        28   
107  Played  4405654        28   
108  Played  4405654        28   
109  Played  4405654        28   
110  Played  4405654        27   
111  Played  4405654        27   
112  Played  4405654        27   
113  Played  4405654        27   
114  Played  4405654        27   
115  Played  4405654        27   
116  Played  4405654       

In [34]:
matches = matches.replace({'home_team':{
        'Manchester United': 'Man Utd',
        'Tottenham Hotspur': 'Spurs',
        'West Ham United': 'West Ham',
        'Manchester City': 'Man City',
        'Brighton & Hove Albion': 'Brighton',
        'Stoke City': 'Stoke',
        'AFC Bournemouth': 'Bournemouth',
        'West Bromwich Albion': 'West Brom',
        'Leicester City': 'Leicester',
        'Swansea City': 'Swansea',
        'Huddersfield Town': 'Huddersfield',
        'Newcastle United': 'Newcastle'
    }, 'away_team': {
        'Manchester United': 'Man Utd',
        'Tottenham Hotspur': 'Spurs',
        'West Ham United': 'West Ham',
        'Manchester City': 'Man City',
        'Brighton & Hove Albion': 'Brighton',
        'Stoke City': 'Stoke',
        'AFC Bournemouth': 'Bournemouth',
        'West Bromwich Albion': 'West Brom',
        'Leicester City': 'Leicester',
        'Swansea City': 'Swansea',
        'Huddersfield Town': 'Huddersfield',
        'Newcastle United': 'Newcastle'
    }}
)

In [ ]:
matches.drop(columns=['label'], inplace=True)
matches['match'] = matches['home_team'] + '-' + matches['away_team']

In [ ]:
defence['match'] = defence['home_team'] + '-' + defence['away_team']

In [ ]:
match_def = pd.merge(right=matches, left=defence, right_on='match', left_on='match')

In [38]:
match_def.columns

Index(['match_id', 'home_team_x', 'away_team_x', 'match_date',
       'home_team_defense', 'away_team_defense', 'home_RB', 'home_R-CB',
       'home_L-CB', 'home_LB', 'home_RCB', 'home_CB', 'home_LCB', 'home_RWB',
       'home_LWB', 'away_RB', 'away_R-CB', 'away_L-CB', 'away_LB', 'away_RCB',
       'away_CB', 'away_LCB', 'away_RWB', 'away_LWB', 'home_backline',
       'away_backline', 'match', 'status', 'roundId', 'gameweek', 'teamsData',
       'seasonId', 'dateutc', 'winner', 'venue', 'wyId', 'date', 'referees',
       'duration', 'competitionId', 'score', 'home_team_y', 'away_team_y'],
      dtype='object')

In [ ]:
match_def.drop(columns=[
    'status', 'roundId', 'seasonId', 'date', 'match_date', 'winner',
    'match_id', 'duration', 'competitionId', 'home_team_y', 'away_team_y'
], inplace=True)

In [40]:
match_def.to_pickle('../data/matches_defence.pkl')

# Event Data Pre-Processing

## Normalizing Names 

In [ ]:
events_com['firstName'] = events_com['firstName'].swifter.set_npartitions(
    8).apply(lambda x: x.encode().decode('unicode_escape').replace('\xad',''))
events_com['lastName'] = events_com['lastName'].swifter.set_npartitions(
    8).apply(lambda x: x.encode().decode('unicode_escape').replace('\xad',''))

In [ ]:
players['firstName'] = players['firstName'].swifter.set_npartitions(
    8).apply(lambda x: x.encode().decode('unicode_escape').replace('\xad',''))
players['lastName'] = players['lastName'].swifter.set_npartitions(
    8).apply(lambda x: x.encode().decode('unicode_escape').replace('\xad',''))
players['playerName'] = players['firstName']+players['lastName']
players['playerName'] = players['playerName'].apply(lambda x: unidecode(x))

In [57]:
players['playerName'] = players['playerName'].apply(lambda x: x.replace('-', ''))

In [141]:
players['playerName'] = players['playerName'].apply(lambda x: x.replace(' ', ''))

In [142]:
players.to_pickle('../data/players.pkl')

## Modifying match_def 

In [92]:
match_def.columns

Index(['home_team_x', 'away_team_x', 'home_team_defense', 'away_team_defense',
       'home_RB', 'home_R-CB', 'home_L-CB', 'home_LB', 'home_RCB', 'home_CB',
       'home_LCB', 'home_RWB', 'home_LWB', 'away_RB', 'away_R-CB', 'away_L-CB',
       'away_LB', 'away_RCB', 'away_CB', 'away_LCB', 'away_RWB', 'away_LWB',
       'home_backline', 'away_backline', 'match', 'gameweek', 'teamsData',
       'dateutc', 'venue', 'wyId', 'referees', 'score', 'home_footedness'],
      dtype='object')

In [107]:
match_def_away = match_def[['wyId', 'away_team_x', 'away_team_defense', 'away_RB', 'away_R-CB', 'away_L-CB', 'away_LB', 'away_RCB',
       'away_CB', 'away_LCB', 'away_RWB', 'away_LWB', 'away_backline', 'match', 'gameweek', 'teamsData',
       'dateutc', 'venue', 'referees', 'score']]

In [108]:
match_def_away.columns = ['wyId', 'team', 'team_defense', 'RB', 'R-CB', 'L-CB', 'LB', 'RCB',
       'CB', 'LCB', 'RWB', 'LWB', 'backline', 'match', 'gameweek', 'teamsData',
       'dateutc', 'venue', 'referees', 'score']

In [110]:
match_def = match_def[['wyId', 'home_team_x', 'home_team_defense', 'home_RB', 'home_R-CB', 'home_L-CB', 'home_LB', 'home_RCB',
       'home_CB', 'home_LCB', 'home_RWB', 'home_LWB', 'home_backline', 'match', 'gameweek', 'teamsData',
       'dateutc', 'venue', 'referees', 'score']]

In [111]:
match_def.columns = ['wyId', 'team', 'team_defense', 'RB', 'R-CB', 'L-CB', 'LB', 'RCB',
       'CB', 'LCB', 'RWB', 'LWB', 'backline', 'match', 'gameweek', 'teamsData',
       'dateutc', 'venue', 'referees', 'score']

In [115]:
match_def = match_def.append(match_def_away, ignore_index=True)

In [119]:
match_def.sort_values(by=['gameweek'], inplace=True, ascending=False)

In [121]:
match_def.reset_index(drop=True)

wyId            team  \
0    2500089         Burnley   
1    2500090  Crystal Palace   
2    2500098         Everton   
3    2500097       Leicester   
4    2500096           Stoke   
5    2500095        Man City   
6    2500094         Chelsea   
7    2500093         Watford   
8    2500092        Brighton   
9    2500091         Arsenal   
10   2500090       West Brom   
11   2500089     Bournemouth   
12   2500094       Newcastle   
13   2500096         Swansea   
14   2500091    Huddersfield   
15   2500092       Liverpool   
16   2500093         Man Utd   
17   2500095     Southampton   
18   2500098        West Ham   
19   2500097           Spurs   
20   2500080         Arsenal   
21   2500083     Southampton   
22   2500088           Spurs   
23   2500087       Newcastle   
24   2500084        West Ham   
25   2500079         Swansea   
26   2500086  Crystal Palace   
27   2500081         Man Utd   
28   2500082       Liverpool   
29   2500080         Burnley   
30   2500085    Huddersfield   
31   2500081        Brighton   
32   2500085        Man City   
33   2500082         Chelsea   
34   2500087         Watford   
35   2500086           Stoke   
36   2500083         Everton   
37   2500088       West Brom   
38   2500084       Leicester   
39   2500079     Bournemouth   
40   2500072           Stoke   
41   2500075     Bournemouth   
42   2500070  Crystal Palace   
43   2500069         Burnley   
44   2500072       Liverpool   
45   2500073         Man Utd   
46   2500078        West Ham   
47   2500077           Spurs   
48   2500076         Chelsea   
49   2500073         Arsenal   
50   2500076         Swansea   
51   2500070       Leicester   
52   2500071         Everton   
53   2500069        Brighton   
54   2500077         Watford   
55   2500078        Man City   
56   2500071    Huddersfield   
57   2500075     Southampton   
58   2500074       West Brom   
59   2500074       Newcastle   
60   2500059     Bournemouth   
61   2500061           Spurs   
62   2500059         Man Utd   
63   2500064     Southampton   
64   2500066           Stoke   
65   2500065        Man City   
66   2500068       West Brom   
67   2500067         Watford   
68   2500062         Chelsea   
69   2500064       Leicester   
70   2500061        Brighton   
71   2500065         Swansea   
72   2500063       Newcastle   
73   2500066         Burnley   
74   2500060         Arsenal   
75   2500068       Liverpool   
76   2500067  Crystal Palace   
77   2500063         Everton   
78   2500062    Huddersfield   
79   2500060        West Ham   
80   2500054         Arsenal   
81   2500058           Stoke   
82   2500053       West Brom   
83   2500055         Chelsea   
84   2500049       Leicester   
85   2500052     Bournemouth   
86   2500056         Everton   
87   2500057        Man City   
88   2500051    Huddersfield   
89   2500050  Crystal Palace   
90   2500052       Liverpool   
91   2500056         Swansea   
92   2500050        Brighton   
93   2500058        West Ham   
94   2500057           Spurs   
95   2500054       Newcastle   
96   2500053         Man Utd   
97   2500055     Southampton   
98   2500049         Burnley   
99   2500051         Watford   
100  2500043         Everton   
101  2500040     Southampton   
102  2500040         Arsenal   
103  2500041        Brighton   
104  2500044       Leicester   
105  2500046           Stoke   
106  2500047         Watford   
107  2500048       West Brom   
108  2500045        Man City   
109  2500042         Chelsea   
110  2500045         Man Utd   
111  2500042        West Ham   
112  2500048         Swansea   
113  2500039     Bournemouth   
114  2500046           Spurs   
115  2500044       Newcastle   
116  2500041    Huddersfield   
117  2500039  Crystal Palace   
118  2500043       Liverpool   
119  2500047         Burnley   
120  2500034         Swansea   
121  2500035       Newcastle   
122  2500032  Crystal Palace   
123  2500031         Chelsea   
124  250

In [122]:
match_def.to_pickle('../data/match+def_lineup.pkl')

## Adding Footedness 

In [3]:
match_def = pd.read_pickle('../data/matches/match+def_lineup.pkl')

In [4]:
a = match_def[['team_defense', 'backline']].explode('team_defense')

In [3]:
players = pd.read_pickle('../data/players.pkl')

In [ ]:
p = players[['playerName', 'foot']]

In [ ]:
b = pd.merge(right=p, left=a, right_on='playerName', left_on='team_defense', how='left')

In [162]:
b[b['foot'].isna()]['team_defense'].unique()

array(['RamiroFunesMori', 'KurtZouma', 'Danilo', 'CesarAzpilicueta',
       'EzequielSchelotto', 'GaetanBong', 'HectorBellerin', 'AhmedHegazi',
       'JamaalLascelles', 'AngelRangel', 'Zanka', 'ChrisLwe',
       'EricBailly', 'MarcosRojo', 'CdricSoares', 'AngeloOgbonna',
       'HctorBellern', 'DavinsonSanchez', 'JavierManquillo', 'TommySmith',
       'Bruno', 'GatanBong', 'NicolsOtamendi', 'CsarAzpilicueta',
       'AntonioRdiger', 'JosHolebas', 'SamusColeman', 'AllanRomoNyom',
       'NathanAk', 'JosephGomez', 'AlbertoMoreno', 'LuisAntonioValencia',
       'VictorLindelf', 'DavinsonSnchez', 'NicolasOtamendi',
       'NachoMonreal', 'FedericoFernndez', 'SebastianPrdl',
       'CedricSoares', 'JoelMatip', 'MiguelBritos', 'VictorLindelof',
       'JamesCollins', 'CucoMartina', 'DavidLuiz', 'MollaWagu',
       'JrmyPied', 'ChancelMbemba', 'PabloZabaleta', 'KikoFemenia',
       'CheikhouKouyat', 'KikoFemena', 'JoseFonte', 'JosFonte',
       'JesusGamez'], dtype=object)

In [65]:
b.loc[b['team_defense'].isin([
    'CesarAzpilicueta',
    'CsarAzpilicueta',
    'HectorBellerin',
    'HctorBellern',
    'EricBailly',
    'KurtZouma',
    'Danilo',
    'EzequielSchelotto',
    'AhmedHegazi',
    'JamaalLascelles',
    'AngelRangel',
    'Zanka',
    'CdricSoares',
    'CedricSoares',
    'DavinsonSanchez', 
    'JavierManquillo',
    'TommySmith',
    'NicolsOtamendi',
    'NicolasOtamendi',
    'AntonioRdiger',
    'AntonioRudiger',
    'SamusColeman',
    'AllanRomoNyom',
    'JosephGomez',
    'LuisAntonioValencia',
    'VictorLindelf',
    'VictorLindelof',
    'DavinsonSnchez',
    'FedericoFernndez',
    'SebastianPrdl',
    'JoelMatip',
    'JamesCollins',
    'DavidLuiz',
    'CucoMartina',
    'MollaWagu',
    'JoseFonte',
    'JosFonte',
    'JrmyPied',
    'ChancelMbemba',
    'PabloZabaleta',
    'KikoFemenia',
    'CheikhouKouyat',
    'KikoFemena',
    'Bruno',
    'JesusGamez'
]), 'foot'] ='right'

In [66]:
b.loc[b['team_defense'].isin([
    'MarcosRojo',
    'RamiroFunesMori',
    'GaetanBong',
    'GatanBong',
    'ChrisLwe',
    'AngeloOgbonna', 
    'JosHolebas',
    'NathanAk',
    'AlbertoMoreno',
    'NachoMonreal',
    'MiguelBritos'    
]), 'foot'] = 'left'

In [67]:
b['foot'].isna().sum()

0

In [68]:
b.index = a.index

In [ ]:
f = [list(b['foot'][i]) for i in a.index.unique()]
f = pd.Series(f)
match_def['footedness'] = f.values
match_def['footedness'] = match_def['footedness'].apply(lambda x: '-'.join(x))

In [195]:
match_def.reset_index(drop=True, inplace=True)
match_def.rename(columns={'wyId': 'matchId'})

In [72]:
match_def.to_pickle('../data/matches/match+def_lineup+footedness_ver2.pkl')

In [ ]:
events_com['playerName'] = events_com['firstName']+events_com['lastName']
events_com.drop(columns=[
    'currentTeamId', 'passportArea', 'weight', 'firstName', 'lastName',
    'middleName', 'birthDate', 'height', 'role', 'birthArea', 'shortName',
    'currentNationalTeamId'
], inplace=True)
events_com.rename(columns={'id': 'eventId', 'wyId': 'playerId'})

In [201]:
events_com.to_pickle('../data/events+player.pkl')

## Mapping Event Tags

In [53]:
event = pd.read_pickle('../data/events+player.pkl')

In [ ]:
tag_key = pd.read_csv('../data/events/tags2name.csv')

In [ ]:
k = list(zip(tag_key['Tag'], tag_key['Description']))

In [50]:
def unpack(x):
    x = [list(x[i].values()) for i in range(0, len(x))]
    return x

In [61]:
def functag2name(x):
    tag_names = list()
    for val in x:
        val = int(val)
        val_pair = [item for item in k if item[0]==val]
        tag_names.append(val_pair[0][1])
    return tag_names

In [ ]:
event['tags'] = event['tags'].swifter.set_npartitions(8).apply(lambda x: unpack(x))
event['tags'] = event['tags'].swifter.set_npartitions(8).apply(lambda x: list(chain(*x)))

In [ ]:
event['tags'] = event['tags'].swifter.set_npartitions(8).apply(lambda x: functag2name(x))

In [63]:
event.to_pickle('../data/events+players.pkl')

## Converting Positions Co-ordinates 

In [49]:
events = pd.read_pickle('../data/events+players.pkl')

In [33]:
def clean_coordinates(x):
    x = [list(d.values()) for d in x]
    x = [l[::-1] for l in x]
    return x

In [34]:
events['positions'] = events['positions'].swifter.set_npartitions(8).apply(lambda x: clean_coordinates(x))

In [ ]:
events['positions'] = events['positions'].swifter.set_npartitions(8).apply(lambda x: [np.multiply([-1.04, 0.68], i) for i in x])
events['positions'] = events['positions'].swifter.set_npartitions(8).apply(lambda x: [np.subtract([0, 68], i) for i in x])

In [62]:
events['playerName'] = events["playerName"].apply(lambda x: unidecode(x))
events['playerName'] = events["playerName"].apply(lambda x: x.replace(' ',''))
events['playerName'] = events["playerName"].apply(lambda x: x.replace('-',''))

In [63]:
def roundcoords(x):
    final_roundedcoords = list()
    for l in x:
        roundedcoords = [np.round(val,decimals=2) for val in l]
        final_roundedcoords.append(roundedcoords)
    return final_roundedcoords

In [64]:
events['positions'] = events['positions'].apply(lambda x: roundcoords(x))

In [65]:
df_players = pd.read_pickle('../data/players.pkl')

In [66]:
# Adding Roles
roles_temp = df_players['role'].values

roles = list()
for i in roles_temp:
    roles.append(i['code3'])
    

players_roles = list(zip(roles,df_players['wyId'],df_players['playerName']))

df_players_roles = pd.DataFrame(players_roles,columns = ['role','playerId','playerName1'])

df_events_roles = events.merge(df_players_roles, left_on = 'playerId', right_on = 'playerId'  )

df_events_roles.drop(['playerName1'], axis = 1, inplace = True)

df_events_roles.to_pickle('../data/events/events_v2.pkl')

# Result

1. Use **events_plot-a-match.pkl** to plot match events from same match simultaneously<br><br>
2. Use **events_v2.pkl** to plot match events from different matches cumulatively over a season - used in all the analysis<br><br>
3. Use **match+def_lineup+footedness_ver2.pkl** for Defensive lineups of teams in each match from the PL 17-18 season<br><br>
**Note:** All event coordinate data has origin on the bottom left corner of the pitch when viewed horizontally and x = [0,104], y= [0,68]